### ISSUE TITLE
Option to disable unattended updates per-app

### ISSUE THREAD

**Issue by luke-jr** [accrescent/accrescent#545]
Unattended updates are not always desirable, but it makes sense to leave it enabled by default. It should be possible for the user to decide per-app to disable (or enable if default is disabled?) unattended updates.

Additionally, it should be possible for an app to disable unattended updates by default. For example, it is important in some cases that nobody (even the developer) can remotely update the app.

Even when unattended upgrades are enabled, it might also be desirable for a new update to prompt the user for agreement to new terms or warning about removal of a feature (etc). The user might also want to switch to another branch (eg, if a LTS option is available)

Let me know if I should split these three up into separate issues.

**Comment by lberrymage** [comment]
I'll try to address these one-by-one.

I'm open to adding an option to disable automatic updates per-app. However, I'd like to know what use cases you or other people have for this.

Could you elaborate on the second feature you're proposing? Are you suggesting the app developer should be able to set a flag so their app isn't automatically updated by default, or something else?

Promoting the user on updates could possibly be useful I think, thought how and when this would occur would need to be defined. For example, it doesn't seem practical to *enforce* notifying the user of feature removals, and they could presumably already do this through a future changelog / "what's new" feature. Is there something in particular you were suggesting?

"Branches" are planned to be supported in the form of separate beta and stable channels for each app. This system could be expanded in the future to accommodate for other use cases such as LTS versions (although this paradigm seems to be very uncommon on the Android ecosystem).

**Comment by luke-jr** [comment]
>I'm open to adding an option to disable automatic updates per-app. However, I'd like to know what use cases you or other people have for this.

It changes a "trust the vendor forever" assumption to a "trust the vendor as they are today".

>Could you elaborate on the second feature you're proposing? Are you suggesting the app developer should be able to set a flag so their app isn't automatically updated by default, or something else?

Yes, an option for the app developer. For example, it is important for Bitcoin's decentralisation that nobody has the ability to remotely update node software; that the user chooses to do so explicitly. This is especially true when an update includes protocol changes.

>Promoting the user on updates could possibly be useful I think, thought how and when this would occur would need to be defined. For example, it doesn't seem practical to enforce notifying the user of feature removals, and they could presumably already do this through a future changelog / "what's new" feature. Is there something in particular you were suggesting?

For example, when the Taproot softfork was rolled out for Bitcoin, I configured the Ubuntu PPAs to require explicit consent to the protocol change, along with a brief description and link to websites to learn more.

**Comment by lberrymage** [comment]
> It changes a "trust the vendor forever" assumption to a "trust the vendor as they are today".

Fair enough. I think a per-app toggle would make sense then, especially since we already have a global toggle.

> Yes, an option for the app developer. For example, it is important for Bitcoin's decentralisation that nobody has the ability to remotely update node software; that the user chooses to do so explicitly. This is especially true when an update includes protocol changes.
>
> For example, when the Taproot softfork was rolled out for Bitcoin, I configured the Ubuntu PPAs to require explicit consent to the protocol change, along with a brief description and link to websites to learn more.

I see. That makes sense for very decentralized systems like Bitcoin as you're describing, but the Android app ecosystem is quite different, so I'm finding it difficult to believe that would really be useful or commonly desired by developers in this context. And I'm hesitant to add it for the seemingly very few who want it because of the potential consequences of users not receiving timely app updates and using old, possibly vulnerable software.

Maybe requiring the user to opt-in to respecting such explicit consent markers would solve both problems since the user would be aware that automatic updates may pause any time. I still have doubts though that this sort of feature is desired by many developers.

**Comment by luke-jr** [comment]
Perhaps even with automatic updates disabled, making it prominent that an update is available and needs manual interaction?

**Comment by lberrymage** [comment]
Possibly. Could you make a separate issue for this though since it's separate from a per-app automatic updates toggle?

**Comment by ghost** [comment]
As another voice: I have no use case for unattended updates. I do want updates to the latest security patches as soon as they are available! I do update regularly. But I am a different case because I use Accrescent from a profile that is kept at rest much of the time. I dislike these programs running in the background or sitting with their data in AFB for the profile. Otherwise I would only not want unattended updates for apps that have new changelog updates.

I prefer to read changelogs before installing apps, not after. I don't care about generic "fixed bugs and improved stability." Waking up to see that my apps updated overnight and lost features or otherwise provide less value to myself is undesirable.

Also, a reason to disable per-app unattended updates: you no longer want to update that app at all or just for a period of time. Maybe a malicious takeover happens. It takes you a few weeks to confirm it, but my accomplice updates it first and tells me that its bad. Maybe I see a post on the developer's website that a third party library in the app has a new vulnerability and asks people to skip the update? How could Accrescent handle these cases? Keys get stolen. Laptops get hacked.
